ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/decoders/TomlAbstractDecoder.kt

```diff
@@ -75,36 +80,70 @@ public abstract class TomlAbstractDecoder : AbstractDecoder() {
         }
     }
 
+    private inline fun <reified T> decodeFloatingPoint(content: Double, lineNo: Int): T = when (T::class) {
+        Float::class -> validateAndConvertFloatingPoint(content, lineNo, FLOAT) { num: Double -> num.toFloat() as T }
+        Double::class -> validateAndConvertFloatingPoint(content, lineNo, DOUBLE) { num: Double -> num as T }
+        else -> invalidType(T::class.toString(), "Signed Type")
+    }
+
+    /**
+     * ktoml parser treats all integer literals as Long and all floating-point literals as Double,
+     * so here we should be checking that there is no overflow with smaller types like Byte, Short and Int.
+     */
+    private inline fun <reified T> validateAndConvertFloatingPoint(
+        content: Double,
+        lineNo: Int,
+        limits: FloatingPointLimitsEnum,
+        conversion: (Double) -> T,
+    ): T = if (content in limits.min..limits.max) {
```

Comment:
what's about suggestion to convert to lower type and then back and check that value is the same?

Reply:
@nulls , such suggestions valid, however looks a little bit confusing for integer types. But for floating-point types it can cause an error. 

Floating-point arithmetics in Kotlin compiler is partially implemented using C++ interop. And in C++ standard it is explicitly told that floating point numbers cannot store all values exactly (not due to limits, but due to precision), so there can potentially be cases when `doubleNumber.toFloat().toDouble() != doubleNumber` due to loss of bits. 

So it is a big question for me if I really should rely on such hack and have potential (but with low probability) case when we corrupt the number and throw an exception with a higher complexity instead of simple check of bounds.

As you know, I am not a risky guy, so I think that I would better use low risk option here (at least with floating point numbers).